File: 백준/Gold/2470. 두 용액/두 용액.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] nums = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .sorted()
                .toArray();

        int left = 0;
        int right = n-1;
        int bestL = 0;
        int bestR = n-1;
        int minDiff = Math.abs(nums[left] + nums[right]);

        while (left < right) {
            int sum = nums[left] + nums[right];
            int currentDiff = Math.abs(sum);
            if (currentDiff < minDiff) {
                minDiff = currentDiff;
                bestL = left;
                bestR = right;
            }

            if (sum == 0) {
                break;
            } else if (sum < 0) {
                left++;
            } else {
                right--;
            }
        }

        System.out.println(nums[bestL] + " " + nums[bestR]);
    }
}

File: 백준/Gold/2470. 두 용액/두 용액.kt
import kotlin.math.abs

/**
 * p2470 두 용액 (투포인터)
 */
fun main() {
    val n = readln().toInt()
    val nums = readln().split(" ").map { it.toInt() }.sorted()

    var left = 0
    var bestLeft = 0
    var right = n-1
    var bestRight = n-1
    var minDiff = abs(nums[left] + nums[right])

    while (left < right) {
        val sum = nums[left] + nums[right]
        val currentDiff = abs(sum)
        if (currentDiff < minDiff) {
            minDiff = currentDiff
            bestLeft = left
            bestRight = right
        }

        if (sum == 0) break
        else if (sum < 0) left++
        else right--
    }

    println("${nums[bestLeft]} ${nums[bestRight]}")
}